src/legacy_http.c

```diff
@@ -340,9 +340,9 @@ static int fill_buffer(HTTP_FILE *file, size_t want, CURLM* multi_handle)
  *
  * Removes `want` bytes from the front of the buffer.
  */
-static int use_buffer(HTTP_FILE *file, int want)
+static int use_buffer(HTTP_FILE *file, size_t want)
 {
-    if((file->buffer_pos - want) <= 0){
+    if(file->buffer_pos <= want){
```

Comment:
What's the type of `buffer_pos`?

You should use a fixed 64-bit type like `uint64_t`, or `unsigned long long`. On 32-bit architectures, `size_t` is 4 bytes in size.

Reply:
`buffer_pos` is `size_t`: https://github.com/AppImage/zsync2/blob/86cfd3a1d6a27483ec40edd62c1a6bd409cbbe5d/src/legacy_http.c#L78

Reply:
All these need to have a real 8-byte size type like `uint64_t`. I'm very certain `size_t`  will be too small on 32-bit, same goes for `off_t`. I just tried with `gcc -m32`:

```c
#include <stdio.h>
#include <sys/types.h>

int main() {
    printf("%lu\n", sizeof(off_t));
}
```

It's generally a good idea to use fixed-size types, as they're predictable and not compiler-dependent. And in this case, it's even a must, since we also have 32-bit binaries. You can't support big files on 64-bit only.

Reply:
That is, by the way, the big issue with the patch in #31.

Reply:
It appears that just opening a 2GB+ file with a 32-bit build is currently broken, e.g. at

https://github.com/AppImage/zsync2/blob/86cfd3a1d6a27483ec40edd62c1a6bd409cbbe5d/src/zsclient.cpp#L986

so just increasing type size on those systems won't help.

Reply:
Please define "broken".

In any case, it's not an excuse for using types with a too small size.

Reply:
> Please define "broken".

Broken as in the file can't be opened:

```
Checking for changes...
Error opening file $file-larger-than-2GB
```

>  it's not an excuse for using types with a too small size.

The types are the same size as - or larger than - the current ones. This change just aligns types within the code to the structs and is one step towards better support for large files.

I don't understand why fixing (or at least improving) 64-bit builds is blocked by 32-bit builds that are currently broken anyway.

Reply:
It would be nice if this was at least fixed in 64-bit builds. I was trying to use zsync2 on files of 100G when I created #31, as it looked like the right tool for the job.

Reply:
Could be. But I don't see the point in accepting a pull request that doesn't completely solve an issue, while it claims to do so. Renaming the PR doesn't make things better, really.

Reply:
The PR was renamed specifically because you pointed out that it doesn't actually solve the whole issue, yet does still improve things for 64-bit builds (i.e. a 64-bit build can download files larger than 2GB, which, for me at least, makes things better than when #31 was opened in 2018).

Reply:
The argument is nonsense. There is no reason not to just swap the types you change _anyway_ for ones that actually work on 32-bit as well. It doesn't have to make 32-bit work entirely... Nobody requested this.

But your PR is really off-putting in that its argument is "at least it's better than nothing". Of course, a full-featured 32-bit fix would be nice. But I'm certain you won't be willing to work on that. At least you could fix the types in the lines you touch anyway, permanently.

Reply:
> It doesn't have to make 32-bit work entirely... Nobody requested this

I may have misinterpreted this statement:

> I don't see the point in accepting a pull request that doesn't completely solve an issue

Reply:
My point is, if you know an exact set of lines has to be touched in the future because the data types don't work on one platform properly, it really makes sense to just use the future proof data type. I basically had to re-do your changes to change the types to the future proof ones. This resulted in *twice* the work overall.